### PR TITLE
feat: implement actual password changing

### PR DESCRIPTION
**Pull request by danielgent**
what is this called? it's called resetting the password but it's actually changing it...  Reset to me implies going back to 0... *anyway...*

_this does actually work!_ :partying_face: 

#### Next
* I could write tests for all this *BUT*
* I think I should investigate email verification on Auth0 *and* come up with a solution. It's the last piece of the puzzle

### CODE REVIEW

File: src/email-templates/password-reset.mjml
Comment (danielgent) [inline src/email-templates/password-reset.mjml:1]: this was copy-pasted from the magic link email which confused me

File: src/routes/tsoa/dbconnections.ts [hunk 1]
```diff
@@ -185,6 +185,8 @@ export class DbConnectionsController extends Controller {
 
     const state = session.id;
 
+    await env.data.universalLoginSessions.create(session);
```
Comment (danielgent) [inline src/routes/tsoa/dbconnections.ts:188]: lol. Obviously the session needed persisting!

File: src/routes/tsoa/dbconnections.ts [hunk 2]
```diff
@@ -160,15 +160,7 @@ export class DbConnectionsController extends Controller {
...
-      // rest copy pasted from authorize route...
-      // redirect_uri,
-      // scope,
-      // state,
-      // audience,
-      // nonce,
-      // response_type,
-      // code_challenge,
-      // code_challenge_method,
+      username: email,
```
Comment (danielgent) [inline src/routes/tsoa/dbconnections.ts:163]: this is the only other thing we need!

File: src/routes/tsoa/login.ts [hunk 3]
```diff
@@ -530,7 +530,7 @@ export class LoginController extends Controller {
   ): Promise<string> {
     const { env } = request.ctx;
     const session = await env.data.universalLoginSessions.get(state);
-    if (!session?.username) {
+    if (!session) {
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:533]: ah, maybe I should revert this. This error wasn't quite correct and threw me off

File: src/routes/tsoa/login.ts [hunk 4]
```diff
@@ -543,11 +543,15 @@ export class LoginController extends Controller {
       throw new HTTPException(400, { message: "Client not found" });
     }
 
-    // TODO - filter by primary user
-    const [user] = await env.data.users.getByEmail(
-      client.tenant_id,
-      session.authParams.username,
-    );
+    // Note! we don't use the primary user here. Something to be careful of
+    // this means the primary user could have a totally different email address
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:547]: I'll worry about this later :smile:

File: src/routes/tsoa/login.ts [hunk 5]
```diff
@@ -561,7 +565,7 @@ export class LoginController extends Controller {
       }
 
       await env.data.passwords.update(client.tenant_id, {
-        user_id: session.authParams.username,
+        user_id: user.id,
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:568]: another tricky existing bug :smile:

File: src/templates/universal/reset-password.ts [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
     <span class="login100-form-title p-b-49">
-        Login with Code
+        Change password
     </span>
     <div class="wrap-input100 validate-input m-b-23">
-        <span class="label-input100">Username</span>
-        <input class="input100" type="text" name="username" placeholder="Type your username" value="{{username}}"
+        <span class="label-input100">New password - at some point we'll add a second confirmation one AND we need to tell the user it
+        has to have whatever rules we specify... OR we just echo back error messages from auth2 in HTML</span>
+        <input class="input100" type="text" name="password" placeholder="Type your password" value="{{password}}"
```
Comment (danielgent) [inline src/templates/universal/reset-password.ts:9]: this is something to think of.  an info message telling the user the rules, and then error messages server-side rendered telling the users which rules they've broken

I don't know how far we want to go with this... the register user page on login2 needs similar functionality but in React we can respond to the keystrokes easily

We could always revert to jQuery here, or even HTMX, or we just echo it back from the server like the good o' days (we probably get some kind of error now)

File: src/email-templates/password-reset.mjml [hunk 7]
```diff
@@ -7,136 +7,59 @@
     </mj-attributes>
     <mj-style>
       .custom-table table {
-        width: auto !important;
-        margin: 0 auto !important;
+      width: auto !important;
```
Comment (markusahlstrand) [inline src/email-templates/password-reset.mjml:10]: Interesting :) Is it an prettier update or did we never format this before?

File: src/routes/tsoa/dbconnections.ts (hunk 1 above)
Comment (markusahlstrand) [reply src/routes/tsoa/dbconnections.ts:180]: Ohh.. interesting.. Happy if this works, but untested territory

File: src/templates/universal/reset-password.ts (hunk 6 above)
Comment (markusahlstrand) [reply src/templates/universal/reset-password.ts:9]: I have htmx on my todo-list of things to investigate :) Haven't really grasped what it's doing. Heard good things about vanilla-js as well..

**Review by markusahlstrand — APPROVED** [review]
Looks good! Think the universal auth stuff might need to tests

**Comment by danielgent** [comment]
> Looks good! Think the universal auth stuff might need to tests

I can frame this where you're asking me for testing :smile: 

*yes* this all needs testing indeed. I want to get email validation working first though so I know this route actually works. It really should all work!

I'd love to actually test these HTML rendered routes but when I tried previously there was some reason I couldn't... ah, we couldn't actually enter the text into the web page... TBD anyway!

File: src/routes/tsoa/dbconnections.ts (hunk 1 above)
Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:180]: Anything concerning? Was this not the flow you were envisaging?

I'm not actually sure why you have this `state` param being passed around but I'm getting there's some CSRF protection happening

File: src/templates/universal/reset-password.ts (hunk 6 above)
Comment (danielgent) [reply src/templates/universal/reset-password.ts:9]: Now the browser wars are over we could do things in vanilla JS as we don't need polyfilling

**Pull request merged**